File: HackerRank Solutions/Dynamic Programming/Unbounded KnapSack.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {

    private static int UKS (int value[], int n, int k) {
		int Max[] = new int[k + 1], temp = -9999;
		Max[0] = 0;
		for (int i = 1; i <= k; i++) {
			Max[i] = Max[i - 1];
			for (int j = 0; j < n; j++) {
				if (i >= value[j]) {
					temp = Max[i - value[j]] + value[j];
				}
				if (temp > Max[i]) {
					Max[i] = temp;
				}
			}
		}
		return Max[k];
	}

	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int T;
		T = in.nextInt();
		while ((T--) > 0) {
			int n = in.nextInt(), k = in.nextInt();
			int value[] = new int[n];
			for (int i = 0; i < n; i++) {
				value[i] = in.nextInt();
			}
			System.out.println(UKS(value, n, k));
		}
		in.close();
	}
    
}

File: HackerRank Solutions/Dynamic Programming/Unbounded Knap Sack.cpp
#include <bits/stdc++.h>

using namespace std;

int UnboundedKnapSack (int value[], int n, int k) {
	int Max[k + 1], temp;
	for (int i = 0; i <= k; i++) {
		Max[i] = 0;
	}
	for (int i = 1; i <= k; i++) {
		Max[i] = Max[i - 1];
		for (int j = 0; j < n; j++) {
			if (i >= value[j]) {
				temp = Max[i - value[j]] + value[j];
			}
			if (temp > Max[i]) {
				Max[i] = temp;
			}
		}
	}
	return Max[k];
}

int main (void) {
	int T, n, k;
	cin >> T;
	while (T--) {
		cin >> n >> k;
		int value[n];
		for (int i = 0; i < n; i++) {
			cin >> value[i];
		}
		cout << UnboundedKnapSack (value, n, k) << endl;
	}
	return 0;
}
